### ISSUE TITLE
Would you be willing to share your opinion on a Whisper-encoder VAD?

### ISSUE THREAD

**Issue by grider-transwithai** [CrispStrobe/CrispASR#36]
Hi @CrispStrobe,

Really impressive work on CrispASR — unifying 18 ASR backends in one C++ binary with no Python deps is genuinely useful, and the care in the pipeline shows.

I'm not here to advertise anything. I just wanted an honest opinion from someone who's done serious engineering on audio AI, and your work made you the obvious person to ask.

I'm the author of TransWithAI/Whisper-Vad-EncDec-ASMR-onnx (https://huggingface.co/TransWithAI/Whisper-Vad-EncDec-ASMR-onnx), a small VAD on top of a frozen `whisper-base` encoder. It was trained on ~500 hours of Japanese ASMR, but in my own testing, it seems to handle normal English, Chinese, and Japanese speech reasonably well too. The problem is I don't have enough out-of-domain data to actually verify that — "works on my clips" isn't a real claim.

If you ever have a spare moment, would you be willing to try it on some of the audio you already work with and tell me what you think? Negative impressions are just as valuable to me as positive ones — I'd rather know where it falls apart than oversell it.

Completely understand if it's not a good use of your time. Thanks either way for the project.

**Comment by CrispStrobe** [comment]
i would recommend you run your model through the run_eval of https://github.com/huggingface/open_asr_leaderboard if you want objective metrics

**Comment by grider-transwithai** [comment]
Hi @CrispStrobe, thanks for the suggestion!

I looked into the Open ASR Leaderboard, but it seems to only evaluate full speech-to-text transcription (WER/RTFx). Since my model is strictly a standalone VAD, run_eval wouldn't give me isolated VAD metrics without coupling it to an ASR pipeline first.

If you happen to know of any good VAD-specific benchmarks, I'd love to hear about them. If not, no worries at all—thanks again for your time and the great work on CrispASR!

**Comment by CrispStrobe** [comment]
Sorry for the confusion earlier. For isolated VAD metrics (DER, precision/recall, false alarm), `pyannote.metrics` gives you the evaluation framework — it's just a scoring library, language doesn't matter. Like so:
```
from pyannote.metrics.detection import DetectionErrorRate
metric = DetectionErrorRate()

for file in test_set:
    speech = your_pipeline(file)          # any VAD model's output
    _ = metric(
        file['annotation'],               # reference
        speech,                           # hypothesis
        uem=file['annotated'])            # region to evaluate

detection_error_rate = abs(metric)
```

For Japanese test data specifically, CALLHOME Japanese (telephone speech) is available on HuggingFace (`talkbank/callhome`, `jpn` split) with annotations. It's a harder domain than ASMR, which is the point. CSJ or ReazonSpeech would likely cover more varied Japanese (broadcast, spontaneous) but require more setup to extract clean VAD labels. 

Any way you would have to load the audio, load the RTTM annotations, convert to pyannote.core.Annotation objects, then feed into the same metric.

Architecture-wise, using a frozen whisper-base encoder for VAD is interesting but heavy. Silero VAD (LSTM-based, ~4M params) and FireRedVAD (DFSMN, ~588K params) make for strong contenders — though I haven't been able to independently verify FireRedVAD's reported numbers yet as the benchmark details don't appear to be publicly released. One angle that could make your approach worthwhile: if the ASR pipeline already has Whisper loaded, the encoder is already in memory, so the VAD head adds negligible overhead.

**Comment by grider-transwithai** [comment]
> Sorry for the confusion earlier. For isolated VAD metrics (DER, precision/recall, false alarm), `pyannote.metrics` gives you the evaluation framework — it's just a scoring library, language doesn't matter. Like so:
> 
> ```
> from pyannote.metrics.detection import DetectionErrorRate
> metric = DetectionErrorRate()
> 
> for file in test_set:
>     speech = your_pipeline(file)          # any VAD model's output
>     _ = metric(
>         file['annotation'],               # reference
>         speech,                           # hypothesis
>         uem=file['annotated'])            # region to evaluate
> 
> detection_error_rate = abs(metric)
> ```
> 
> For Japanese test data specifically, CALLHOME Japanese (telephone speech) is available on HuggingFace (`talkbank/callhome`, `jpn` split) with annotations. It's a harder domain than ASMR, which is the point. CSJ or ReazonSpeech would likely cover more varied Japanese (broadcast, spontaneous) but require more setup to extract clean VAD labels.
> 
> Any way you would have to load the audio, load the RTTM annotations, convert to pyannote.core.Annotation objects, then feed into the same metric.
> 
> Architecture-wise, using a frozen whisper-base encoder for VAD is interesting but heavy. Silero VAD (LSTM-based, ~4M params) and FireRedVAD (DFSMN, ~588K params) make for strong contenders — though I haven't been able to independently verify FireRedVAD's reported numbers yet as the benchmark details don't appear to be publicly released. One angle that could make your approach worthwhile: if the ASR pipeline already has Whisper loaded, the encoder is already in memory, so the VAD head adds negligible overhead.

Thanks for the recommendations, I will run some standard benchmarks you suggested. Also, you make a really fair point of using whisper from the ASR stage, though my current trained model is on the base model, and I assume most people would use large v2 and v3, but I will experiment with that. Anyway, thanks a lot for your suggestions, I will try bring some numbers back, including FireRedVAD.

**Comment by CrispStrobe** [comment]
looking at your model, it seems the encoder was fine-tuned and not frozen though?

**Comment by himekifee** [comment]
AFAI can tell, it was [frozen](https://github.com/TransWithAI/whisper-vad/blob/c5ea2819bb0020504d0a2eee9df2344e94860f60/configs/encoder_decoder_base.yaml#L6).

**Comment by CrispStrobe** [comment]
it is now usable per crispasr with `--vad-model whisper-vad-asmr-q4_k.gguf`

Indeed in the training config there is freeze_encoder: true, but weight comparison shows the attention layers were fine-tuned (Q/K/V/O projections differ by 0.3–0.7 from whisper-base). Only the FFN layers appear frozen (diffs ~0.001, consistent with float16 rounding).

This means the encoder output cannot be reused directly for whisper ASR decoding — the representations have diverged from the original whisper-base encoder. The VAD model works correctly as a standalone VAD, but encoder sharing is not possible.

**Comment by grider-transwithai** [comment]
Thanks again for the concrete benchmark suggestions.

I ran a local benchmark pass with `vad-bench`, using isolated VAD scoring via `pyannote.metrics` rather than an ASR-coupled WER/RTFx leaderboard metric.

My takeaway is cautious: the open Whisper-VAD-ASMR ONNX model is benchmarkable and has some useful recall-oriented behaviour, but it does not look clearly strong enough to position as production-ready from these results. I would keep it framed as research/evaluation-only for now.

## Setup

- Model under test: public/open TransWithAI Whisper-VAD-ASMR ONNX artefact (`TransWithAI/Whisper-Vad-EncDec-ASMR-onnx`)
- Local ONNX metadata: `openai/whisper-base`, 20 ms frame output, 30 s chunks, ONNX opset 17
- Primary dataset: CALLHOME Japanese via `talkbank/callhome` / CALLHOME Japanese public mirror, 120 recordings
- Additional sanity check: AMI IHM 20 short chunks
- Baselines: Silero VAD and FireRedVAD
- Metrics: `pyannote.metrics`, `collar=0.0`, reporting DER / false alarm / miss / precision / recall / F1 / detection cost

## CALLHOME Japanese full benchmark

This is the main result I would treat as relevant here, since it is Japanese telephone speech and is meaningfully out-of-domain relative to the ASMR training setting.

- Recordings: 120
- Total evaluated duration: 67,468.760 s
- Reference speech duration: 62,418.378 s
- Collar: 0.0

| Model | DER ↓ | F1 ↑ | Precision ↑ | Recall ↑ | Detection cost ↓ | False alarm s ↓ | Miss s ↓ |
|---|---:|---:|---:|---:|---:|---:|---:|
| FireRedVAD | **0.107404** | **0.944814** | 0.971657 | 0.919415 | 0.143303 | 1,673.980 | 5,029.998 |
| Whisper-VAD-ASMR ONNX | 0.111760 | 0.943756 | 0.949952 | **0.937640** | 0.199403 | 3,083.416 | **3,892.434** |
| Silero VAD | 0.115088 | 0.939822 | **0.984906** | 0.898685 | **0.118541** | **859.668** | 6,323.926 |

The result is mixed rather than a clear win for Whisper-VAD-ASMR. FireRedVAD is best on DER/F1. Silero is best on precision, detection cost, and false-alarm duration. Whisper-VAD-ASMR is best only on recall and miss duration, which suggests it is more recall-oriented in this setup, but at the cost of more false alarms and worse detection cost.

## AMI IHM 20-chunk sanity check

I also ran a small AMI IHM chunked check on CUDA. I would not treat this as a replacement for CALLHOME or a full AMI meeting benchmark, because this manifest is only 20 short speech-heavy chunks and does not stress false-alarm behaviour realistically.

- Clips: 20
- Total duration: 25.34 s
- Collar: 0.0

| Model | DER ↓ | F1 ↑ | Precision ↑ | Recall ↑ | Detection cost ↓ | False alarm s ↓ | Miss s ↓ |
|---|---:|---:|---:|---:|---:|---:|---:|
| Whisper-VAD-ASMR ONNX/CUDA | **0.279400** | **0.837615** | **1.000000** | **0.720600** | **0.209550** | **0.000** | **7.080** |
| FireRedVAD/CUDA | 0.303473 | 0.821121 | 1.000000 | 0.696527 | 0.227605 | 0.000 | 7.690 |
| Silero VAD/CUDA | 0.556038 | 0.614922 | 1.000000 | 0.443962 | 0.417028 | 0.000 | 14.090 |

Whisper-VAD-ASMR leads most non-tied metrics on this small AMI check, but I would not put much weight on that result: precision and false alarms are tied because the chunks are speech-heavy, and the sample is too small to say much about real meeting behaviour.

## Caveats

- I would keep the Whisper-VAD-ASMR model framed as research/evaluation-only for now, not production-ready.
- CALLHOME Japanese is the primary out-of-domain check here; the AMI result is only a small chunk-level sanity check.
- The FireRedVAD numbers are from my local run under the same evaluation harness; I am not treating them as independent verification of any unpublished upstream benchmark claims.
- Detection cost and DER rank models differently, so I would report both rather than saying one model is simply “better.”
- Given your later weight-comparison finding, I am treating Whisper-VAD-ASMR as a standalone VAD result. I am not claiming that its encoder output can be reused directly for Whisper ASR decoding.
- I would still want production-specific validation before any deployment claim: streaming behaviour, latency, memory/CPU/GPU profiling, threshold tuning, noisy/long-form audio, and integration testing in CrispASR.

Overall, I would position this as: the open ONNX VAD path is useful to evaluate, but the current numbers do not make a strong production case. On the more relevant CALLHOME Japanese benchmark, Whisper-VAD-ASMR mainly shows a recall/miss advantage, while FireRedVAD and Silero are stronger on other important metrics. I would keep it in the research/evaluation bucket unless further tuning and broader validation improve the tradeoff.
